### PR TITLE
BibFormat: recjson update using BibReformat

### LINKED ISSUE #1708: BibUpload: update recjson after metadata update
Originally on 2014-02-19

Currently, "recjson" is not being updated after a record gets modified via a bibupload job.  This can lead to discrepancy in metadata between MARCXML and recjson.

BibUpload should either update recjson, or else delete it to prevent this discrepancy, exactly like it is doing for "recjstruct" now.  

Whether to update or delete is governed by configuration variable `CFG_BIBUPLOAD_SERIALIZE_RECORD_STRUCTURE`.

A test case could be plugged into bibupload regression test suite to make sure the updates are fine.

### PR DESCRIPTION
**Pull request by egabancho**
- Adds `recjson` to the list of available formats that BibReformat
  could use. (closes #1708)

Signed-off-by: Esteban J. G. Gabancho esteban.gabancho@gmail.com

### CODE REVIEW

**Comment by drjova** [comment]
:+1: lgtm

File: modules/miscutil/lib/upgrades/invenio_2014_11_04_format_recjson.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    return "New format recjson to format table."
+
+
+def do_upgrade():
+    """Upgrade recipe procedure."""
+    run_sql("ALTER TABLE format MODIFY code varchar(10) NOT NULL")
+    if not run_sql("SELECT id FROM format WHERE code='recjson'"):
+        run_sql("INSERT INTO format "
+                "(name,code,description,content_type,visibility) "
+                "VALUES ('recjson','recjson', 'recjson record represetation', "
```
Comment (tiborsimko) [inline modules/miscutil/lib/upgrades/invenio_2014_11_04_format_recjson.py:38]: "represetation" is a typo

File: modules/miscutil/sql/tabfill.sql [hunk 2]
```diff
@@ -310,6 +310,7 @@ INSERT INTO format (id,name,code,description,content_type,visibility) VALUES (26
 INSERT INTO format (id,name,code,description,content_type,visibility) VALUES (27,'Mobile brief','mobb', 'Mobile brief format.', 'text/html', 0);
 INSERT INTO format (id,name,code,description,content_type,visibility) VALUES (28,'Mobile detailed','mobd', 'Mobile detailed format.', 'text/html', 0);
 INSERT INTO format (id,name,code,description,content_type,visibility) VALUES (29,'WebAuthorProfile data helper','wapdat', 'cPickled dicts', 'text', 0);
+INSERT INTO format (id,name,code,description,content_type,visibility) VALUES (30,'recjson','recjson', 'recjson record represetation', 'application/json', 0);
```
Comment (tiborsimko) [inline modules/miscutil/sql/tabfill.sql:313]: s/represetation/representation

File: modules/bibformat/doc/admin/bibformat-admin-guide.webdoc [hunk 3]
```diff
@@ -885,14 +885,14 @@ in order to decrease the load of your server. To do so, use the <code>bibreforma
...
+ -o, --format            Specify output format to be (re-)created. (default HB and recjson)
  -n, --noprocess         Count records to be processed only (no processing done)
  Example: bibreformat -n Show how many records are to be bibreformated.
 </pre>
 </blockquote>
-For example, to reformat all records in HB (=HTML brief) format, you'd launch:
+For example, to reformat all records in HB (=HTML brief) and recjson formats, you'd launch:
 <blockquote>
-<pre>$ bibreformat -a -oHB
+<pre>$ bibreformat -a -oHB,recjson
```
Comment (tiborsimko) [inline modules/bibformat/doc/admin/bibformat-admin-guide.webdoc:895]: It is a bit weird to have one format in uppercase, one in lowercase.  Could they use the same case?

File: modules/bibformat/lib/bibreformat.py [hunk 4]
```diff
@@ -59,13 +64,16 @@ def store_last_updated(fmt, iso_date):
...
     @param fmt: output format to use
-    @param sql: dictionary with pre-created sql queries for various cases (for selecting records). Some of these queries will be picked depending on the case
-    @param sql_queries: a list of sql queries to be executed to select records to reformat.
-    @param cds_query: a search query to be executed to select records to reformat
+    @param sql: dictionary with pre-created sql queries for various cases (for
+        selecting records). Some of these queries will be picked depending on
+        the case
+    @param sql_queries: a list of sql queries to be executed to select records
+        to reformat.
+    @param cds_query: a search query to be executed to select records to
```
Comment (tiborsimko) [inline modules/bibformat/lib/bibreformat.py:75]: Parameters like `cds_query` (and more) don't exist anymore.  If you update docstring, can you also clean away the ones that are gone?

File: modules/bibformat/lib/bibreformat.py [hunk 5]
```diff
@@ -198,33 +206,60 @@ def check_validity_input_formats(input_formats):
...
 ### (see iterate_over_old further down)
 
-def iterate_over_new(recIDs, fmt):
+
+def _update_recjson_format(recid, *args, **kwargs):
+    """Update recjson cache.
+
+    :param int recid: record id to process
+    """
+    get_record(recid, reset_cache=True)
```
Comment (tiborsimko) [inline modules/bibformat/lib/bibreformat.py:215]: `dummy = get_record(...)` for kwalitee aesthetics?

File: modules/miscutil/lib/upgrades/invenio_2014_11_04_format_recjson.py (hunk 1 above)
Comment (egabancho) [inline modules/miscutil/lib/upgrades/invenio_2014_11_04_format_recjson.py:38]: Done!

File: modules/miscutil/sql/tabfill.sql (hunk 2 above)
Comment (egabancho) [inline modules/miscutil/sql/tabfill.sql:313]: and done!

File: modules/bibformat/lib/bibreformat.py [hunk 6]
```diff
@@ -198,33 +206,60 @@ def check_validity_input_formats(input_formats):
...
+                                                    on_the_fly=True,
+                                                    save_missing=False)
+    save_preformatted_record(recID=recid,
+                             of=fmt,
+                             res=record,
+                             needs_2nd_pass=needs_2nd_pass,
+                             low_priority=True)
+
+
+CFG_BIBFORMAT_UPDATE_FORMAT_FUNCTIONS = {'recjson': _update_recjson_format}
```
Comment (tiborsimko) [inline modules/bibformat/lib/bibreformat.py:235]: People would expect a `CFG_...` to live either in global `invenio.conf`, or in local `bibformat_config.py`.  Perhasp we should move it to the latter?  Or call it `_CFG_...` with leading underscore to indicate that this is a very internal beast?

File: modules/bibformat/lib/bibreformat.py [hunk 7]
```diff
@@ -370,7 +408,7 @@ def main():
 Option -c prevents from finding records in private collections.
 
 Examples:
-  bibreformat                    Format all new or modified records (in HB).
+  bibreformat                    Format all new or modified records (in HB and recjson).
```
Comment (tiborsimko) [inline modules/bibformat/lib/bibreformat.py:411]: Again, RECJSON in uppercase for consistency?

**Pull request merged**